### DISCUSSION TITLE
Server behavior is different than normal way

### DISCUSSION THREAD

**Discussion by Elikill58** [symfony/symfony#58037] [Q&A] [answered]
I'm adding test to my project. I encouter multiple issue, but the most important one is about the strange way that phpunit load the symfony website.

For example, in normal env, the variable `$_SERVER["CONTEXT_DOCUMENT_ROOT"]` exists, but not with test env.
Also, there is no session when doing request, which means we can't same something on session (as user does when they login), or even get the token manager.

When using `$client->getContainer()->get("security.csrf.token_manager")->getToken("thing")` we have a message as "Currently no session available".

Why is there so much differences between both, and not simply the same website just load with `ENV=test` ?

**Comment by stof** [comment]
When running tests with WebTestCase, the KernelBrowser of browserkit handles a Request with the Kernel directly. It does not send an actual HTTP request to an actual HTTP server. And so PHP superglobals are not populated.

Regarding the session, where are you trying to read the CSRF token ? Is it during the handling of a request or in your test outside the request handling ? Not having a session available outside the request handling is totally expected, as the session is associated to the request.

**Reply by Elikill58** [reply]
Ok, this leads to different behavior but maybe you can't do better.

I'm trying to create one CSRF token to make a request to my rest API. So, it's before making the request, here is my code:
```php
$client = static::createClient();
$client->loginUser(new LoginUser("TestUser"));
$client->request('POST', "/test", [
    "_csrf_token" => $this->getCsrfToken("tester") // here I want the csrf token
]);
```

**Reply by MatTheCat** [reply]
The same way `KernelBrowser::loginUser` seeds a session containing a security token, you can seed one containing your CSRF token. The key would be `SessionTokenStorage::SESSION_NAMESPACE.'/'.$tokenId`, and the value the token you’ll use in you test.

**Reply by Elikill58** [reply]
@MatTheCat how? The method `getCsrfToken` that I used in example code is own method that return static token, it doesn't do anything

**Reply by stof** [reply]
@Elikill58 and that's precisely the issue. A static token will not be a valid token on the server.
A test submitting a static value is basically a test trying to perform a CSRF attack (and then, failing is a good output for that test)

**Reply by Elikill58** [reply]
@stof I know. That's why I changed the backend check, to add this static token when env is test.

I cannot add test for valid token as I can NOT GENERATE new token on test. I can't test the right way, with valid token, except with my patch composed of "fake" token.

**Reply by MatTheCat** [reply]
I’m using the following traits in my functional tests:

```php
trait SessionTrait
{
    protected static function seedSession(KernelBrowser $client, array $data): void
    {
        $sessionFactory = $client->getContainer()->get('session.factory');
        \assert($sessionFactory instanceof SessionFactoryInterface);
        $session = $sessionFactory->createSession();
        $session->replace($data);
        $session->save();
        $client->getCookieJar()->set(new Cookie($session->getName(), $session->getId()));
    }
}
```

```php
trait CsrfTokenTrait
{
    protected static function getCsrfTokenSeed(string $tokenId, string $token): array
    {
        return [SessionTokenStorage::SESSION_NAMESPACE."/{$tokenId}" => $token];
    }
}
```

Then, after the client is created I can seed a CSRF token with

```php
self::seedSession($client, self::getCsrfTokenSeed('token_id', 'token'));
```

and pass `token` in the request.

**Reply by Elikill58** [reply]
Thanks! It seems good code, but as I login user before, it unlog it or if I log if after, it failed to find the token

**Reply by MatTheCat** [reply]
Indeed you’d need a trait to seed the session with a security token as well. Here is mine from the same project for reference; it targets [symfonycorp/connect](https://github.com/symfonycorp/connect) so you would need to adapt it though.

```php
trait SymfonyConnectLoginTrait
{
    public static function getLoginSeed(User $user, array $roles = ['ROLE_USER'], string $firewall = 'main'): array
    {
        return ['_security_'.$firewall => serialize(new ConnectToken(
            new InMemoryUser($user->getUuid(), null, $roles),
            'access token',
            $user,
            $firewall,
            'SCOPE_EMAIL SCOPE_LOCATION SCOPE_PUBLIC',
            $roles,
        ))];
    }
}
```
Then you’d merge `getLoginSeed` and `getCsrfTokenSeed` return and pass it to `seedSession`.

**Reply by Elikill58** [reply, marked as answer]
Perfect, it works fine ! here is my final code:
```java
public function fillCsrfToken(KernelBrowser $client, string $tokenId): string {
    $generated = rtrim(strtr(base64_encode(random_bytes(32)), '+/', '-_'), '=');
    $this->seedSession($client, $this->getCsrfTokenSeed($tokenId, $generated));
    return $generated;
}

private function getCsrfTokenSeed(string $tokenId, string $token): array {
    return [SessionTokenStorage::SESSION_NAMESPACE."/{$tokenId}" => $token,
        '_security_main' => serialize(new PostAuthenticationToken(new LoginUser("MyUser"), "main", ["ROLE_USER"]))
    ];
}

private function seedSession(KernelBrowser $client, array $data): void {
    $sessionFactory = $client->getContainer()->get('session.factory');
    \assert($sessionFactory instanceof SessionFactoryInterface);
    $session = $sessionFactory->createSession();
    $session->replace($data);
    $session->save();
    $client->getCookieJar()->set(new Cookie($session->getName(), $session->getId()));
}
```
Then I use like this:
```java
$client = self::getLoggedClient();
$token = $this->fillCsrfToken($client, "tester");
$client->request('POST', "/test", [
    "_csrf_token" => $token
]);
```
